### ISSUE TITLE
Feature Request - "locking" hidden taskbar notification icon locations

### ISSUE THREAD

**Issue by moccor** [ramensoftware/windhawk-mods#1170]
There is a weird name for that little arrow on the taskbar, but when mousing over it, it says "show hidden icons". Basically the feature I think would be convenient and fit with many of the other great taskbar Windhawk mods, is if there was a way to "lock" the positions of icons in that area. They can be moved around (at least on W11), but they always kinda just seem to be random, likely whatever order they load in at Windows startup must decide what order they are placed in, but only for that specific OS state. Once shutdown/rebooted, one or more icons can be in different locations.

It isn't a big inconvenience, but it sounds like a cool and efficient feature that can fit in with many of the already very efficient taskbar mods. If more info is needed, feel free to ask

**Comment by Marshall-Brooks** [comment]
I would like to see this also. There was a payware software call PS Tray Factory that could do this back in the Win XP days, but I don't know if it is still around - also allowed you to show or hide systray icons.

**Comment by m417z** [comment]
From my experience, tray icon order is preserved, both on the taskbar and in the overflow popup. That is unless these are several icons of the same program, in which case I believe they'll be added one after the other, ordered by the time they're created.

**Comment by Marshall-Brooks** [comment]
Actually, I tend to agree. I tend to see the icons in the same order. Maybe Win11 (or previous?) added a registry key to preserve the order.

I'm sure WinXP and previous did not, hence the usefulness of PS Tray Factory.

Show/Hide icons would still be useful - mainly b/c there are a few programs that I don't need to see in the systray, but lack an option to hide the notification icon, unless there is a registry Key for it.

Found some info here:
https://superuser.com/questions/1635898/show-hide-specific-icon-on-windows-notification-area-with-command-line

In HKCU\Control Panel\NotifyIconSettings, there is a registry key for UIOrderList. It is binary, but I'll bet that stores the tray order.

According to the link, underNotifyIconSettings there will be a bunch of folders with numbers and under each subfolder, there will be a key for "IsPromoted", however, looking through the keys, I found keys that I wanted to remove from the systray that had the icon I wanted to remove, had IsPromoted() set to 0, but where not removed.

The syspin program also did not seem to work for me.

**Comment by Marshall-Brooks** [comment]
Did a bit more research on this ...

Google isn't much help b/c Microsoft considers "hidden" as "in the systray overflow area, whereas I consider "hidden" as "Not shown at all".

I believe the isPromoted registry key changes whether the icon is in overflow or not, but does not hide the icon.

I found a post saying PSTrayFactory does work in Windows 11, although the official page doesn't show it as supported.

I found an alternative called TrayControl - which is supposed to do this as well - but for me, it didn't work. It could hide Taskbar apps, but the systray panel was blank, even though it looked like it should work.

According to this: https://superuser.com/questions/1707389/hide-some-system-tray-icons-uncollapsed - it could be done with an AutoHotKey script, but I have AHK 2.0 and the script was written for AHK 1.0 and they aren't backwards compatible. I tried to update it, but it didn't work for me.

**Comment by moccor** [comment]
I agree that it is often preserved, but for some reason it does randomly become disorganized, I am not sure what causes it. I know I organized mine once and then they got mixed up. I will see how long it will last like this again, maybe some kind of Windows updates screw with it.

**Comment by Marshall-Brooks** [comment]
Sorry for basically hi-jacking the thread.

Follow-up question ... The AHK script is supposed to give you the window handle of the icon, and then when the icon is loaded, you run a different script that hides the icon, but as I said, it has to be converted to work with AHK 2.0, and even then, I'm not sure if it works with Win11.

I used UWPSpy and figured out the icon property, for example:
![image](https://github.com/user-attachments/assets/5e17415b-2041-4931-94f1-2491b11d5e02)
Above is the identification for Jabra Direct for my headsets. I want this running to let me know if the software/firmware can be updated, but I don't need the icon always in the systray.

Is there a way to hide this icon using one of the existing Windhawk extensions - perhaps Taskbar Styler?

Thanks in advance!!!

**Comment by Marshall-Brooks** [comment]
Confirmed a bit ...

I found out you don't have to convert AHK scripts to AHK 2.0. You can just change the extension to .ah1 and open them with AutoHotKeyU64.exe.

I also found out the script does not work with my version of Windows 11. It runs put it doesn't show anything in the window - similar to the results I got with TrayControl.

I'm thinking perhaps they used to work before the taskbar became a UWP.
